File: extras/org.eclipse.rwt.widgets.codemirror/src/org/eclipse/rwt/widgets/codemirror/internal/CodeMirror.js
/*
 * polymap.org
 * Copyright 2011-2012, Falko Bräutigam. All rights reserved.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 */

var loadedScripts = new Object();

/**
 * Loads additional javascript in the document.
 * 
 * @param url
 */
function loadScript( url, callback, context ) {
    if (loadedScripts[url] != null) {
        // mimic the deferred callback call
        setTimeout( function() { callback( context ); }, 10 );
        return;
    }
    loadedScripts[url] = url;
    
    var script = document.createElement( "script" );
    script.type = "text/javascript";

    if (callback != null) {
        if (script.readyState) { // IE
            script.onreadystatechange = function() {
                if (script.readyState == "loaded" || script.readyState == "complete") {
                    script.onreadystatechange = null;
                    callback( context );
                }
            };
        } 
        else { // Others
            script.onload = function() {
                callback( context );
            };
        }
    }
    script.src = url;
    document.getElementsByTagName( "head" )[0].appendChild( script );
}

/**
 * Loads additional CSS in the document.
 * 
 * @param url
 */
function loadCSS( url ) {
    if (loadedScripts[url] != null) {
        return;
    }
    loadedScripts[url] = url;

    var ref = document.createElement( "link" );
    ref.setAttribute( "rel", "stylesheet" );
    ref.setAttribute( "type", "text/css" );
    ref.setAttribute( "href", url );
    document.getElementsByTagName("head")[0].appendChild( ref );
}


/**
 * JavaScript of the CodeMirror RWT Widget
 * 
 * @author <a href="http://www.polymap.de">Falko Bräutigam</a>
 */
qx.Class.define( "org.eclipse.rwt.widgets.CodeMirror", {
	extend : qx.ui.layout.CanvasLayout,

	construct : function( id ) {
		this.base( arguments );
		this.setHtmlAttribute( "id", id );
        this.set( { backgroundColor : "white" });
		this._id = id;
        this._codeMirror = null;
        this._libLoaded = false;
        this._lineMarkers = {};
        
        this.initHeight();
        this.initOverflow();
	},

	properties : {
	    text : {
            check : "String",
            init : null,
            apply : "_applyText",
            event : "_applyText"
        }
	},

	members : {
        /**
         * Lazy init function.
         */
	    _init : function( elm ) {
	        var self = this;
	        this._codeMirror = CodeMirror( elm, {
	            value: "text text text...",
	            mode: "text/x-java",
	            theme: "eclipse",
	            indentUnit: 4,
	            lineNumbers: true,
	            matchBrackets: true,
                onChange: function( cm, info ) { self._syncServer( info, false ); },
                onCursorActivity: function( cm, info ) { 
                    self._onFocus( cm, info );
                    self._syncServer( info, false );
                },
                extraKeys: {
                    "Ctrl-S": function( cm ) { self._syncServer( null, true ); }
                }
	        });
	        this._codeMirror.setOption( "theme", "eclipse" );
	        if (this.getText() != null) {
	            this._codeMirror.setValue( this.getText() );
	        }
	    },
	    
	    /**
	     * Called onCursorActivity().
	     */
	    _onFocus : function( cm, info ) {
	        var shell = null;
	        var parent = this.getParent();
	        //alert( parent );
	        while (shell == null && parent != null) {
	            if (parent.classname == "org.eclipse.swt.widgets.Shell") {
	                shell = parent;
	            }
	            parent = parent.getParent();
	        }
	        if (shell != null) {
	            shell.setActiveChild( this );
	        }
	    },
	    
	    /**
	     * 
	     */
	    loadLib : function( lib_url ) {
	        loadCSS( lib_url + "&res=lib/codemirror.css" );
            loadCSS( lib_url + "&res=theme/eclipse.css" );

            loadScript( lib_url + "&res=lib/codemirror.js", function( context ) {
                loadScript( lib_url + "&res=mode/clike/clike.js", function( context ) {
                    qx.ui.core.Widget.flushGlobalQueues();

                    if (!org_eclipse_rap_rwt_EventUtil_suspend) {
                        context._init( document.getElementById( context._id ) );

                        var widgetId = org.eclipse.swt.WidgetManager.getInstance().findIdByWidget( context );
                        var req = org.eclipse.swt.Request.getInstance();
                        req.addParameter( widgetId + ".load_lib_done", "true" );
                        req.send();
                        this._libLoaded = true;
                    }
                }
                , context ); // loadScript
            }
			, this ); // loadScript
		},
		
		/**
		 * Text modifier.
		 *
		 * @type member
		 * @param value {var} Current value
		 * @param old {var} Previous value
		 */
		_applyText : function( value, old ) {
		    //alert( "applyText(): " + this._codeMirror );
		    if (this._codeMirror != null && this._codeMirror.getValue() != value) {
		        this._codeMirror.setValue( value );
		    }
		},
	    
		/**
		 * @param info
		 * @param forceSave
		 */
		_syncServer : function( info, forceSave ) {
		    if (!org_eclipse_rap_rwt_EventUtil_suspend && this._codeMirror != null) {
		        var widgetId = org.eclipse.swt.WidgetManager.getInstance().findIdByWidget( this );
		        var req = org.eclipse.swt.Request.getInstance();
		        var sendDelay = 1500;

		        // text
		        if (this._codeMirror.getValue() != this.getText()) {
		            this.setText( this._codeMirror.getValue() );
		            req.addParameter( widgetId + ".text", this.getText() );
		        }
		        // cursorPos
                var cursorPos = this._codeMirror.getCursor( true );
                req.addParameter( widgetId + ".cursorpos", this._codeMirror.indexFromPos( cursorPos ) );
                // save
                if (forceSave) {
                    req.addParameter( widgetId + ".save", "true" );
                    sendDelay = 0;
                }
                
                if (this.sendTimeout) {
                    clearTimeout( this.sendTimeout );
                }
                this.sendTimeout = setTimeout( function() {
                    // XXX check if server side has a listener
                    req.send();
                }, sendDelay );
		    }
		},
		
		/**
		 * 
		 * @param id (String)
		 * @param line (String/int)
         * @param charStart (String/int)
         * @param charEnd (String/int)
		 */
        setLineMarker: function( id, line, charStart, charEnd, textClassName, text ) {
            if (this._codeMirror) {
                if (this._lineMarkers[id] != null) {
                    clearMarker( this._lineMarkers[id] );
                }
                var marker = this._codeMirror.setMarker( parseInt( line )-1, text );
                marker.id = id;
                
                marker.startPos = this._codeMirror.posFromIndex( parseInt( charStart ) );
                marker.endPos = this._codeMirror.posFromIndex( parseInt( charEnd ) );
                if (marker.startPos.ch != marker.endPos.ch) {
                    marker.markedText = this._codeMirror.markText( marker.startPos, marker.endPos, textClassName );
                }
                
                this._lineMarkers[id] = marker;
            }
        },
        
        clearLineMarker: function( id ) {
            var marker = this._lineMarkers[id];
            if (marker) {
                if (marker.markedText) {
                    marker.markedText.clear();
                    // XXX hack: the above method does not always clear everything
                    this._codeMirror.setLineClass( marker, null, null );
                }
                this._codeMirror.clearMarker( marker );
                this._lineMarkers[id] = null;
            }
        },
        
        clearLineMarkers: function() {
            for (var id in this._lineMarkers) {
                this.clearLineMarker( id );
            }
            this._lineMarkers = {};
        },
        
        setSelection: function( start, end ) {
            var startPos = this._codeMirror.posFromIndex( parseInt( start ) );
            var endPos = this._codeMirror.posFromIndex( parseInt( end ) );
            this._codeMirror.setSelection( startPos, endPos );
        },
        
        openCompletions: function( /**/json ) {
            var completions = eval( json );
            var editor = this._codeMirror;
            
            if (this.complete && this.complete.parentNode) {
                this.complete.parentNode.removeChild( this.complete );
            }
            
            // build the select widget
            var complete = document.createElement("div");
            this.complete = complete;
            complete.className = "CodeMirror-completions";
            var sel = complete.appendChild(document.createElement("select"));
            // Opera doesn't move the selection when pressing up/down in a
            // multi-select, but it does properly support the size property on
            // single-selects, so no multi-select is necessary.
            if (!window.opera) sel.multiple = true;
            for (var i=0; i<completions.length; ++i) {
                var opt = sel.appendChild(document.createElement("option"));
                opt.appendChild(document.createTextNode(completions[i].text));
            }
            sel.firstChild.selected = true;
            sel.size = Math.min(15, completions.length);
            //sel.style.padding = '2px';
            
            var pos = editor.cursorCoords();
            complete.style.left = pos.x + "px";
            complete.style.top = pos.yBot + "px";
            complete.style.zIndex = 200000;
            complete.style.position = 'absolute';
            //var parent = document.getElementById( this._id );
            var parent = document.body;
            parent.appendChild(complete);
            // If we're at the edge of the screen, then we want the menu to appear on the left of the cursor.
            var winW = window.innerWidth || Math.max(document.body.offsetWidth, document.documentElement.offsetWidth);
            if(winW - pos.x < sel.clientWidth)
                complete.style.left = (pos.x - sel.clientWidth) + "px";
            // Hack to hide the scrollbar.
            if (completions.length <= 15)
                complete.style.width = (sel.clientWidth - 1) + "px";

            var done = false;
            function close() {
                if (!done) {
                    done = true;
                    complete.parentNode.removeChild(complete);
                }
            }
            function pick() {
                var completion = completions[sel.selectedIndex];
                var start = editor.posFromIndex( completion.replaceStart );
                var end = editor.posFromIndex( completion.replaceEnd );
                //alert( completion.text + " : start=" + completion.replaceStart + " -> " + start );
                
                editor.replaceRange( completion.text, start, end );
                close();
                setTimeout(function() { editor.focus(); }, 50);
            }
            CodeMirror.connect(sel, "blur", close);
            CodeMirror.connect(sel, "keydown", function(event) {
                var code = event.keyCode;
                // Enter
                if (code == 13) {
                    CodeMirror.e_stop(event); 
                    pick();
                }
                // Escape
                else if (code == 27) {
                    CodeMirror.e_stop(event); 
                    close(); 
                    editor.focus();
                }
                else if (code != 38 && code != 40) {
                    close(); editor.focus();
                    // Pass the event to the CodeMirror instance so that it can handle things
                    // like backspace properly.
                    editor.triggerOnKeyDown(event);
                    //setTimeout(function() { CodeMirror.simpleHint(editor, getHints);}, 50);
                }
            });
            CodeMirror.connect(sel, "dblclick", pick);

            sel.focus();
            // Opera sometimes ignores focusing a freshly created node
            if (window.opera) setTimeout(function(){if (!done) sel.focus();}, 100);
            return true;
        },
        
        executeCode : function( code2eval ) {
		    var self = this;
		    alert( code2eval );
		    window.eval( code2eval );
		}
	}

});

File: extras/org.eclipse.rwt.widgets.codemirror/src/org/eclipse/rwt/widgets/codemirror/CodeMirror.java
/* 
 * polymap.org
 * Copyright (C) 2011-2014, Falko Bräutigam. All rights reserved.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 */
package org.eclipse.rwt.widgets.codemirror;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.TreeMap;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import org.apache.commons.lang.StringUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Control;
import org.eclipse.rwt.widgets.codemirror.internal.CallMethodCommand;
import org.eclipse.rwt.widgets.codemirror.internal.CodeMirrorJSService;
import org.eclipse.rwt.widgets.codemirror.internal.CodeMirrorLCA;
import org.eclipse.rwt.widgets.codemirror.internal.RenderCommand;
import org.eclipse.rwt.widgets.codemirror.internal.SetPropertyCommand;
import org.eclipse.rwt.widgets.codemirror.internal.CodeMirrorLCA.WidgetAdapter;

import org.eclipse.core.runtime.ListenerList;

/**
 * Widget that provides a <a href="http://codemirror.net">CodeMirror</a> syntax
 * highlighting code editor.
 * 
 * @author <a href="http://www.polymap.de">Falko Bräutigam</a>
 */
public class CodeMirror
        extends Control {

    private static Log log = LogFactory.getLog( CodeMirror.class );

    public final static String      PROP_TEXT = "text";
    public final static String      PROP_SELECTION = "selection";
    public final static String      PROP_CURSOR_POS = "cursorpos";
    /** This is a pseudo property that indicated that 'Ctrl-S' was preset on the client. */
    public final static String      PROP_SAVE = "save";
    
    /** External CodeMirror library location. **/
    private String                  js_location   = CodeMirrorJSService.getBaseUrl();

    private String                  text = StringUtils.EMPTY;

    private int                     cursorPos;

    private LineMarkers             lineMarkers = new LineMarkers();

    private TextSelection           selection;
    
    private LCAAdapter              lcaAdapter = new LCAAdapter();
    
    private ListenerList            listeners = new ListenerList();

    
    static {
        CodeMirrorJSService.register();
    }

    
    public CodeMirror( Composite parent, int style ) {
        super( parent, style );
    }

    
    public CodeMirror( Composite parent, int style, String js_location ) {
        this( parent, style );
        this.js_location = js_location;
    }

    
    public Object getAdapter( Class adapter ) {
        if (adapter.isAssignableFrom( CodeMirrorLCA.class )) {
            return new CodeMirrorLCA();
        }        
        else if (adapter.isAssignableFrom( WidgetAdapter.class )) {
            return lcaAdapter;
        }
        else {
            return super.getAdapter( adapter );
        }
    }

    /**
     * XXX Fix IControlThemeAdapter getControlWith() issue when using FormLayout
     */
    @Override
    public int getBorderWidth() {
        return 0;
    }


    public String getJSLocation() {
        return js_location;
    }

//    @Override
//    public void setLayout( final Layout layout ) {
//        throw new UnsupportedOperationException( "setLayout()" );
//    }

    
    public String getText() {
        return text;
    }

    
    public void setText( String text ) {
        this.text = text != null ? text : StringUtils.EMPTY;
        lcaAdapter.pushCommand( new SetPropertyCommand( PROP_TEXT, text ) );
        firePropertyEvent( PROP_TEXT, this.text, null );
    }

    
    public int getCursorPos() {
        return cursorPos;
    }

    
    public LineMarkers lineMarkers() {
        return lineMarkers;    
    }
    
    
    public void setSelection( int start, int end ) {
        selection = new TextSelection( start, end );
        lcaAdapter.pushCommand( new CallMethodCommand( "setSelection", start, end ) );
        firePropertyEvent( PROP_SELECTION, getSelection(), null );        
    }

    public void clearSelection() {
        selection = null;
        lcaAdapter.pushCommand( new CallMethodCommand( "setSelection", 0, 0 ) );
        firePropertyEvent( PROP_SELECTION, getSelection(), null );        
    }
    
    public TextSelection getSelection() {
        return selection;
    }

    /**
     * Add a listener to this editor. The following properties fire events:
     * <ul>
     * <li>{@link #PROP_TEXT}</li>
     * <li>{@link #PROP_SELECTION}</li>
     * <li>{@link #PROP_CURSOR_POS}</li>
     * </ul>
     * 
     * @param l The listener to register.
     */
    public void addPropertyChangeListener( PropertyChangeListener l ) {
        listeners.add( l );
    }

    public void removePropertyCHangeListener( PropertyChangeListener l ) {
        listeners.remove( l );
    }

    protected void firePropertyEvent( String prop, Object newValue, Object oldValue ) {
        PropertyChangeEvent ev = new PropertyChangeEvent( this, prop, oldValue, newValue );
        for (Object l : listeners.getListeners()) {
            ((PropertyChangeListener)l).propertyChange( ev );
        }
    }

    public void openCompletions( List<ICompletion> proposals ) {
        StringBuilder json = new StringBuilder( 1024 );
        json.append( '[' );
        for (ICompletion proposal : proposals) {
            json.append( json.length() > 1 ? "," : "" );
            json.append( "{\"text\":\"" ).append( proposal.getCompletion() ).append( "\"," );
            json.append( "\"replaceStart\":" ).append( proposal.getReplaceStart() ).append( "," );
            json.append( "\"replaceEnd\":" ).append( proposal.getReplaceEnd() ).append( "}" );
        }
        json.append( ']' );
        lcaAdapter.pushCommand( new CallMethodCommand( "openCompletions", json.toString() ) );
    }
    
    /**
     * The internal interface for the {@link CodeMirrorLCA}.
     */
    class LCAAdapter 
            implements WidgetAdapter {

        private boolean                 load_lib_done = false;
        
        private Queue<RenderCommand>    commandQueue = new LinkedList();
        

        public void setText( String text ) {
            String old = CodeMirror.this.text;
            CodeMirror.this.text = text;
            firePropertyEvent( PROP_TEXT, CodeMirror.this.text, old );
        }

        public void setCursorPos( int cursorPos ) {
            int old = CodeMirror.this.cursorPos;
            CodeMirror.this.cursorPos = cursorPos;
            firePropertyEvent( PROP_CURSOR_POS, CodeMirror.this.cursorPos, old );
        }

        public void forceSave() {
            firePropertyEvent( PROP_SAVE, Boolean.TRUE, null );
        }

        public boolean isJSLoaded() {
            return load_lib_done;
        }

        public void setJSLoaded( boolean loaded ) {
            load_lib_done = loaded;
        }

        public Iterable<RenderCommand> nextCommands() {
            Queue<RenderCommand> result = commandQueue;
            commandQueue = new LinkedList();
            return result;
        }
        
        void pushCommand( RenderCommand command ) {
            commandQueue.offer( command );
        }
    };
    
    
    /**
     * 
     */
    public class LineMarkers
            extends TreeMap<Integer,ILineMarker> {

        public void clear() {
            super.clear();
            lcaAdapter.pushCommand( new CallMethodCommand( "clearLineMarkers" ) );
        }

        public ILineMarker put( ILineMarker marker ) {
            return put( marker.getLine(), marker );
        }
        
        @SuppressWarnings("restriction")
        public ILineMarker put( Integer line, ILineMarker marker ) {
            ILineMarker result = super.put( marker.getLine(), marker );

            StringBuilder buf = new StringBuilder( 256 );
            buf.append( "<div style=\"font-weight:bold;display:inline;position:relative;left:0px;top:0px;vertical-align:top;float:left;width:16px;");
//            if (marker.getFgColor() != null) {
//                buf.append( "color:" ).append( marker.getFgColor() ).append( ";" );
//            }
            String textClassName = "cm-error";
            if (marker.getIcon() != null) {
                // XXX insert real image URL here
                String imageUrl = "icons/error_tsk.gif";
                if (marker.getIcon().internalImage.getResourceName().contains( "warn" )) {
                    imageUrl = "icons/warn_tsk.gif";
                    textClassName = "cm-warn";
                }
                else if (marker.getIcon().internalImage.getResourceName().contains( "info" )) {
                    imageUrl = "icons/info_tsk.gif";
                    textClassName = "cm-info";
                }
                buf.append( "background:url(" ).append( js_location ).append( "&res=" + imageUrl + ");" );
            }
            buf.append( "\" " );
            if (marker.getText() != null) {
                buf.append( "title=\"" ).append( marker.getText() ).append( "\"" );
            }
            buf.append( "> </div>%N%" );

            lcaAdapter.pushCommand( new CallMethodCommand( "setLineMarker", 
                    marker.getId(), marker.getLine(), 
                    marker.getCharStart(), marker.getCharEnd(), 
                    textClassName, buf.toString() ) );
            return result;
        }

        public void putAll( Map<? extends Integer, ? extends ILineMarker> map ) {
            throw new RuntimeException( "not yet implemented" );
        }
    }
    
    
    /**
     * 
     */
    public class TextSelection {
        
        private int         start, end;

        protected TextSelection( int start, int end ) {
            this.start = start;
            this.end = end;
        }
        
        public int getStart() {
            return start;
        }
        
        public int getEnd() {
            return end;
        }

        public String getText() {
            return StringUtils.substring( text, start, end );
        }
    }
    
}
